packages/typography/src/Link/Link.tsx

```diff
@@ -21,84 +25,89 @@ import {
+const Link = Polymorphic<LinkProps>(
+  ({
+    href,
+    children,
+    className,
+    arrowAppearance = ArrowAppearance.None,
+    hideExternalIcon = false,
+    baseFontSize: baseFontSizeOverride,
+    target: targetProp,
+    darkMode: darkModeProp,
+    as = 'a' as PolymorphicAs,
+    ...rest
+  }) => {
+    const [currentHostname, setCurrentHostname] = useState('');
+    useEffect(() => {
+      setCurrentHostname(window.location.hostname);
+    }, []);
 
-  const { theme } = useDarkMode(darkModeProp);
+    const { theme } = useDarkMode(darkModeProp);
+    const { Component } = usePolymorphic(as);
 
-  const hrefHostname = useMemo(() => {
-    if (!href) return;
-    const httpRegex = /^http(s)?:\/\//;
-    return httpRegex.test(href) ? new URL(href).hostname : currentHostname;
-  }, [href, currentHostname]);
+    const hrefHostname = useMemo(() => {
+      if (!href) return;
+      const httpRegex = /^http(s)?:\/\//;
+      return httpRegex.test(href) ? new URL(href).hostname : currentHostname;
+    }, [href, currentHostname]);
 
-  const baseFontSize = useUpdatedBaseFontSize(baseFontSizeOverride);
+    const baseFontSize = useUpdatedBaseFontSize(baseFontSizeOverride);
 
-  let target, icon, rel;
+    let target, icon, rel;
 
-  if (targetProp) {
-    target = targetProp;
-  } else {
-    if (hrefHostname === currentHostname) {
-      target = '_self';
+    if (targetProp) {
+      target = targetProp;
     } else {
-      target = '_blank';
-      rel = 'noopener noreferrer';
+      if (hrefHostname === currentHostname) {
+        target = '_self';
+      } else {
+        target = '_blank';
+        rel = 'noopener noreferrer';
+      }
     }
-  }
 
-  if (target === '_blank' && !hideExternalIcon) {
-    icon = (
-      <OpenNewTabIcon role="presentation" className={openInNewTabStyles} />
-    );
-  } else if (arrowAppearance !== ArrowAppearance.None) {
-    icon = (
-      <ArrowRightIcon
-        role="presentation"
-        size={12}
-        className={cx({
-          [arrowRightIconHover]: arrowAppearance === ArrowAppearance.Hover,
-          [arrowRightIconPersist]: arrowAppearance === ArrowAppearance.Persist,
-        })}
-      />
-    );
-  }
+    if (target === '_blank' && !hideExternalIcon) {
+      icon = (
+        <OpenNewTabIcon role="presentation" className={openInNewTabStyles} />
+      );
+    } else if (arrowAppearance !== ArrowAppearance.None) {
+      icon = (
+        <ArrowRightIcon
+          role="presentation"
+          size={12}
+          className={cx({
+            [arrowRightIconHover]: arrowAppearance === ArrowAppearance.Hover,
+            [arrowRightIconPersist]:
+              arrowAppearance === ArrowAppearance.Persist,
+          })}
+        />
+      );
+    }
 
-  const elementProps = href
-    ? ({ as: 'a', href, target, rel } as const)
-    : ({ as: 'span' } as const);
+    const elementProps = href
+      ? ({ as: 'a', href, target, rel } as const)
+      : ({ as: 'span' } as const);
```

Comment:
We shouldn't need to pass in `as` to the `Component` element

Reply:
Should move this logic above the, and call `usePolymorphic(href ? 'as' : 'span')`